File: 0303-range-sum-query-immutable/0303-range-sum-query-immutable.java
class NumArray {

    private int prefixSums[];

    public NumArray(int[] nums) {
        int n = nums.length;

        prefixSums = new int[nums.length];
        prefixSums[0] = nums[0];

        for (int i = 1; i < n; i++) {
            prefixSums[i] = nums[i] + prefixSums[i - 1];
        }    
    }
    
    public int sumRange(int left, int right) {
        return left == 0 ? prefixSums[right] 
                         : prefixSums[right] - prefixSums[left - 1];
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(left,right);
 */

File: 0303-range-sum-query-immutable/0303-range-sum-query-immutable.js
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
    this.dpArray = [nums[0]];
    for (let i = 1; i < nums.length; i++)
        {
            this.dpArray[i] = this.dpArray[i-1] + nums[i];
        }
};

/** 
 * @param {number} left 
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function(left, right) {
    let sum = this.dpArray[right];
    return (left === 0) ? sum : sum - this.dpArray[left-1];
};

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */